# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import plotly.express as px
import seaborn as sns
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [5]:
# Проверим, являются ли пропуски случайными или пропуск в days_employed также имеет пропуск в total_income в одной
# в одной же строке

len(data.loc[(data['days_employed'].isna()) & (data['total_income'].isna())])

2174

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [6]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [7]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [8]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [9]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [10]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [11]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [12]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [13]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [14]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

54

In [16]:
data = data.drop_duplicates()

**Задание 15. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [17]:
data['education'] = data['education'].str.lower()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [18]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [19]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [20]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [21]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [22]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

## Исследуйте данные и ответьте на вопросы

**Задание 19. Есть ли зависимость между количеством детей и возвратом кредита в срок?**

### Создаём функцию `pivot`, вносим в него обработанную и сгруппированную информацию из таблицы `data` чтобы проанализировать нужную выборку данных.  Как результат функции возвращаем сводную таблицу и график.

In [25]:
def pivot(data, column):
    
    """
    Делаем универсальную функцию для создания сводной таблицы где функция принимает датафрейм, группирует, 
    Переименовывает названия колонок. После этого выводим результаты в виде таблицы и графика.
    
    
    count - подсчитает чисто все строки, т.е.  общее число клиентов
    sum - сумму ноликов и единичек, т.е. по факту будет только сумма единиц => кол-во должников
    mean - сумма деленна на кол-во, т.е. те же должники деленные на общее число клиентов
       
    """
    
    pivot = data.pivot_table(index = column, values = 'debt', aggfunc = ['count', 'sum', 'mean']).reset_index()
    pivot.columns = [column, 'Кол-во клиентов', 'Кол-во должников', '% невозврата']
    display(pivot.sort_values('% невозврата', ascending = False))
        
    
    fig = px.bar(pivot, x = column, y = '% невозврата', color = '% невозврата', title = '% невозврата по ' + column )
    fig.show()
    



#### Вызываем функцию `pivot()` с аргументом `data` и`'children'`. Выводим таблицу с количеством заёмщиков, должников и их пропорцией, а также строим график.

In [26]:
pivot(data, 'children')

children  Кол-во клиентов  Кол-во должников  % невозврата
4         4               41                 4      0.097561
2         2             2052               194      0.094542
1         1             4809               444      0.092327
3         3              330                27      0.081818
0         0            14107              1063      0.075353
5         5                9                 0      0.000000

**Вывод:**  Чем больше детей - тем больше долговое бремя у заёмщика, за исключением заёмщиков с тремя детьми у котороых долговая нагрузка несколько ниже чем у заёмщиков с 2-мя и 3-мя детьми. У заёмщиков с 5-ю детьми долги отсутствуют вовсе.

Проблемы:
Результат неярко выраженный и заметного контраста между количеством детей и сроком возврата дога не прослеживется, за исключением заёмщиков с 5-ю детьми у которых долга нет. Однако учитывая что 5-ти детных заёмщиков 9 человек, то результаты можно считать ненадёжными.

Как решить (рекомендации):
Чтобы добиться более представительного результата следует использовать более репрезентативную выбоку. Также, размер выборки, возможно, следует увеличить.

**Задание 20. Есть ли зависимость между семейным положением и возвратом кредита в срок?**

### Вызываем функцию pivot() и применяем в качестве аргумента `data` и `data_family`.

In [28]:
pivot(data, 'family_status')

family_status  Кол-во клиентов  Кол-во должников  % невозврата
0  Не женат / не замужем             2796               273      0.097639
3       гражданский брак             4146               385      0.092861
4        женат / замужем            12266               927      0.075575
1              в разводе             1189                84      0.070648
2         вдовец / вдова              951                63      0.066246

**Вывод:** Зависимость долгового бремени от семейного статуса присутствует.

По долговой нагрузке с отрывом лидируют не состоящие в браке и состоящие в гражданском браке заёмщики. Оставшиеся три категории с наименьшей долговой нагрузкой замыкают вдовы/вдовцы.

**Задание 21. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?**

### Вызываем функцию pivot() и применяем в качестве аргумента `data` и `total_income_category`.

In [30]:
pivot(data, 'total_income_category')

total_income_category  Кол-во клиентов  Кол-во должников  % невозврата
4                     E               22                 2      0.090909
2                     C            15938              1353      0.084891
0                     A               25                 2      0.080000
1                     B             5014               354      0.070602
3                     D              349                21      0.060172

**Вывод:** Корреляции между ежемесячным доходом и задолженностью не просматривается.

**Задание 22. Как разные цели кредита влияют на его возврат в срок?**

### Вызываем функцию pivot() и применяем в качестве аргумента `data` и `purpose_category`.

In [32]:
pivot(data, 'purpose_category')

purpose_category  Кол-во клиентов  Кол-во должников  % невозврата
0    операции с автомобилем             4281               400      0.093436
2     получение образования             3989               369      0.092504
3        проведение свадьбы             2324               183      0.078744
1  операции с недвижимостью            10754               780      0.072531

**Вывод:** Категория кредита влияет на задолженность.

Заметно большую задолженность имеют заёмщики в категориях: операции с автомобилем и получение образования. С отрывом идут займы на свадьбы и операции с недвижимостью. Заёмщики недвижимости самые кредитоспособные.

**Задание 23. Приведите возможные причины появления пропусков в исходных данных.**

*Ответ:* Пропуски могли возникнуть в результате ошибок при привыгрузке данных, из-за неправильного запроса, а также в следствии ошибок при объединении таблиц. 

В исходных данных встречались пропуска в двух колонках: days_employed (2174 пропусков) и total_income (2174 пропусков). Данные пропуски нельзя назвать случайными тк строки с пропущенными days_employed имели одновременно пропуски и в категории total_income. 

После вывдения результата 4-го задания я провожу проверку ниже чтобы убедиться в том что пропусти в days_employed соответствуют пропускам в колрнке total_income.

`len(data.loc[(data['days_employed'].isna()) & (data['total_income'].isna())])`

Возможная причина пропусков это сознательное умалчивание информации. Возможно, у заёмщиков была возможось не включать важную информацию о свох доходах в базу данных, или они изъявили желание не оглашать публично их стаж и уровень дохода.

В результате мы видим что 2174 заёмщиков предоставили не полную информацию которую пришлось заменять медианными значениями чтобы не терять выбоку и провести исследование.

**Задание 24. Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.**

*Ответ:* Пропущенные значения в колонках: days_employed (2174 пропусков) и total_income (2174 пропусков) являются количественными. Для заполнения количественных пропусков предпочтительнее всего использовать характерные значения например: среднее арифметическое или медиану.

В данном исследовании была выбрана медиана чтобы избежать влияния паразитных, оторванных значений на общую картину. Уровень зарплаты и рабочий стаж могут значительно отличатся в пределах выборки, поэтому их влияние следует исключить нахождением медианного значения.

## Общий вывод.

Проведённое исследование позволило выявить ряд зависимостей и определить взаимосвязь социального статуса, доходов и прочих факторов на долговоую нагрузку заёмщиков.

Портрет заёмщика-должника: 
многодетные, состоящие в гражданском браке люди берущие займ на операции с автомобилем или получение образования.

С другой стороны, портрет заёмщика с наименьшей долговой нагрузкой это: бездетные состоящие в официальном браке, вдовы либо вдовцы берущие займ на недвижимость (заёмщики без долга с 5 детьми составляют всего 9 чел поэтому их мы не будем учитывать).

Как ни странно, но явной зависимости между уровнем дохода и задолженностью не наблюдается.

*Рекомендации для будующих исследований*:
В некоторых исследованиях, а именно зависимость количества детей от долговой нагрузки, было показано что заёмщики с 3-мя и 5-ю детьми выбиваются из общей закономерности. Это может быть следствием недостаточно большой или недостаточно репрезентативной выборки. Следовательно, рекомендуется повысить размер и качество выборки для аналогичных исследований в будующем с целью улучшить качество исследования.
